lms/djangoapps/courseware/tests/test_access.py

```diff
@@ -865,3 +865,79 @@ def test_course_catalog_access_num_queries(self, user_attr_name, action, course_
         course_overview = CourseOverview.get_from_id(course.id)
         with self.assertNumQueries(num_queries, table_blacklist=QUERY_COUNT_TABLE_BLACKLIST):
             bool(access.has_access(user, action, course_overview, course_key=course.id))
+
+
+@ddt.ddt
+class CCXTestCase(SharedModuleStoreTestCase):
```

Comment:
This is a somewhat unusual use of `SharedModuleStoreTestCase` because you're not creating anything with `setUpClass`. You're definitely correct to want to only create the course one time. The idea behind `SharedModuleStoreTestCase` is that the modulestore is not re-initialized between test method calls, so you can keep using one or two courses over and over for tests you know are read-only on the course content itself.

But because you're doing your work in a `setUp()` here, you're doing most of that course creation processing anyway (it's going to run for every ddt permutation) and you're just not getting any of the modulestore isolation guarantees. To see how to convert this to something that only sets this data up once, please see the [docstring](https://github.com/edx/edx-platform/blob/d35e88e271bd30cdbc4b8b63c715147582ea17a1/common/lib/xmodule/xmodule/modulestore/tests/django_utils.py#L360-L418) for `SharedModuleStoreTestCase` and its `setupClassAndTestData` method.

Reply:
That makes sense, thanks. Does https://github.com/edx/edx-platform/pull/20746/commits/7b0c8b3acc3c46f16d34074058863c47288acfe9 resolve this? Also I noticed in the docstring: "any Django ORM operations should still happen in setUp()." Not sure whether this means that only the course can go in setUpClass, while the users created should still be in setUp?

Reply:
Ah, so `SharedModuleStoreTestCase` was created before we moved to a Django version that had `setUpTestData`, which is why the kind of awkward `setUpClassAndTestData` exists. The root problem that it's trying to solve is that Django's base test class calls `setUpTestData` from its `setUpClass` method and does all its ORM setup/cleanup there. Meanwhile, all the modulestore initialization is done in `SharedModuleStoreTestCase`'s `setUpClass`.

So that creates a weird situation. In order to get the modulestore isolation guarantees, you normally need to call super(...).setUpClass() at the start of your setUpClass() method before you start making modulestore changes. But that means that `setUpTestData` is running before anything is being created in the modulestore, which is almost never what you want -- courses nearly almost come first, and then you build everything around it, like enrollments and such.

So in the beginning, we told everyone to just do your modulestore setup in `setUpClass` and then do all the Django ORM model creation in `setUp` (so it would be run many times). But if you want to create the Django ORM models only once, while still ensuring proper cleanup, you call `setUpClassAndTestData` in your `setUpClass` and explicitly _don't_ call super(...).setUpClass() directly:

```python

@classmethod
def setUpClass(cls):
    # Note, we are *NOT* calling super(MyTestClass, cls).setUpClass() because
    # setUpClassAndTestData() will do that for us in the right ordering.
    with super(MyTestClass, cls).setUpClassAndTestData():
        # Do all the modulestore operations here, like creating courses
        cls.course = CourseFactory.create(enable_ccx=True)

@classmethod
def setUpTestData(cls):
        ccx = CustomCourseForEdX(
            course_id=cls.course.id,
            coach=coach,
            display_name="Test CCX"
        )
        ccx.save()
        #... etc.
```

As long as all your modulestore data creation is done in the `with super(MyTestClass, cls).setUpClassAndTestData():` block:
* Modulestore data creation will only happen once for the test class and not for each test method.
* Django ORM model creation will only happen once for the test class and not for each test method.
* Everything will be cleaned up properly, both in MongoDB and the Django ORM.
* setUpTestData and all its model creation will only happen after modulestore data has been created.

Reply:
Thanks for the detailed explanation! Hopefully this is fixed now with 1c80b9c334. I'll definitely be saving your comment for future reference. :+1: